File: src/client/java/fr/hardel/asset_editor/client/javafx/lib/data/EnchantmentViewMatchers.java
package fr.hardel.asset_editor.client.javafx.lib.data;

import fr.hardel.asset_editor.client.javafx.lib.data.SlotConfigs.SlotConfig;
import fr.hardel.asset_editor.store.ElementEntry;
import net.minecraft.world.item.enchantment.Enchantment;

import java.util.Locale;
import java.util.Map;
import java.util.Objects;

public final class EnchantmentViewMatchers {

    @FunctionalInterface
    public interface ViewMatcher {
        boolean matches(ElementEntry<Enchantment> entry, String category);
    }

    private static final Map<StudioSidebarView, ViewMatcher> MATCHERS = Map.of(
            StudioSidebarView.SLOTS, EnchantmentViewMatchers::matchesSlot,
            StudioSidebarView.ITEMS, EnchantmentViewMatchers::matchesItem,
            StudioSidebarView.EXCLUSIVE, EnchantmentViewMatchers::matchesExclusive
    );

    public static boolean matches(ElementEntry<Enchantment> entry, String filterPath, StudioSidebarView sidebarView) {
        if (filterPath.isEmpty()) return true;

        String[] parts = filterPath.split("/", 2);
        String category = parts[0];
        String leaf = parts.length == 2 ? parts[1] : "";
        if (!leaf.isEmpty() && !entry.id().getPath().equals(leaf)) return false;

        ViewMatcher matcher = MATCHERS.get(sidebarView);
        return matcher != null && matcher.matches(entry, category);
    }

    public static boolean matchesSlot(ElementEntry<Enchantment> entry, String category) {
        SlotConfig config = SlotConfigs.BY_ID.get(category);
        if (config == null) return false;
        return entry.data().definition().slots().stream()
                .anyMatch(g -> config.slots().contains(g.getSerializedName()));
    }

    public static boolean matchesItem(ElementEntry<Enchantment> entry, String category) {
        boolean supported = entry.data().definition().supportedItems().unwrapKey()
                .map(tag -> tag.location().getPath().equals(category))
                .orElse(false);
        if (supported) return true;

        boolean primary = entry.data().definition().primaryItems()
                .flatMap(hs -> hs.unwrapKey())
                .map(tag -> tag.location().getPath().equals(category))
                .orElse(false);
        if (primary) return true;

        return entry.tags().stream().anyMatch(t -> t.getPath().equals(category));
    }

    private static boolean matchesExclusive(ElementEntry<Enchantment> entry, String category) {
        var tagKey = entry.data().exclusiveSet().unwrapKey();
        if (tagKey.isPresent()) {
            String full = tagKey.get().location().toString().toLowerCase(Locale.ROOT);
            String path = tagKey.get().location().getPath().toLowerCase(Locale.ROOT);
            return full.equals(category) || path.equals(category);
        }
        return entry.data().exclusiveSet().stream()
                .map(holder -> holder.unwrapKey().map(k -> k.identifier()).orElse(null))
                .filter(Objects::nonNull)
                .anyMatch(id -> id.toString().toLowerCase(Locale.ROOT).equals(category)
                        || id.getPath().toLowerCase(Locale.ROOT).equals(category));
    }

    private EnchantmentViewMatchers() {}
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/components/page/enchantment/EnchantmentViewMatchers.kt
package fr.hardel.asset_editor.client.compose.components.page.enchantment

import fr.hardel.asset_editor.client.compose.lib.SlotConfigs
import fr.hardel.asset_editor.client.compose.lib.StudioSidebarView
import fr.hardel.asset_editor.store.ElementEntry
import net.minecraft.world.item.enchantment.Enchantment
import java.util.Locale

object EnchantmentViewMatchers {

    fun interface ViewMatcher {
        fun matches(entry: ElementEntry<Enchantment>, category: String): Boolean
    }

    private val matchers = mapOf(
        StudioSidebarView.SLOTS to ViewMatcher(::matchesSlot),
        StudioSidebarView.ITEMS to ViewMatcher(::matchesItem),
        StudioSidebarView.EXCLUSIVE to ViewMatcher(::matchesExclusive)
    )

    @JvmStatic
    fun matches(entry: ElementEntry<Enchantment>, filterPath: String, sidebarView: StudioSidebarView): Boolean {
        if (filterPath.isEmpty()) {
            return true
        }

        val parts = filterPath.split("/", limit = 2)
        val category = parts[0]
        val leaf = if (parts.size == 2) parts[1] else ""
        if (leaf.isNotEmpty() && entry.id().path != leaf) {
            return false
        }

        val matcher = matchers[sidebarView]
        return matcher != null && matcher.matches(entry, category)
    }

    @JvmStatic
    fun matchesSlot(entry: ElementEntry<Enchantment>, category: String): Boolean {
        return entry.data().definition().slots()
            .any { group -> SlotConfigs.expandsTo(group.serializedName, category) }
    }

    @JvmStatic
    fun matchesItem(entry: ElementEntry<Enchantment>, category: String): Boolean {
        val supported = entry.data().definition().supportedItems().unwrapKey()
            .map { tag -> tag.location().path == category }
            .orElse(false)
        if (supported) {
            return true
        }

        val primary = entry.data().definition().primaryItems()
            .flatMap { holders -> holders.unwrapKey() }
            .map { tag -> tag.location().path == category }
            .orElse(false)
        if (primary) {
            return true
        }

        return entry.tags().stream().anyMatch { tag -> tag.path == category }
    }

    private fun matchesExclusive(entry: ElementEntry<Enchantment>, category: String): Boolean {
        val normalizedCategory = category.lowercase(Locale.ROOT)
        val tagKey = entry.data().exclusiveSet().unwrapKey()
        if (tagKey.isPresent) {
            val location = tagKey.get().location()
            val full = location.toString().lowercase(Locale.ROOT)
            val path = location.path.lowercase(Locale.ROOT)
            return full == normalizedCategory || path == normalizedCategory
        }

        return entry.data().exclusiveSet().stream()
            .map { holder -> holder.unwrapKey().map { key -> key.identifier() }.orElse(null) }
            .filter { id -> id != null }
            .anyMatch { id ->
                id.toString().lowercase(Locale.ROOT) == normalizedCategory ||
                    id.path.lowercase(Locale.ROOT) == normalizedCategory
            }
    }
}